Add faces for syntax elements #2

zonuexe opened the pull request:
refs #350 (Cc: @Fuco1)

This PR keeps appearance compatibility.


## Customized Example

Users will be able to easily match the color of `$` with variables as you like.

### Before (no customize)

<img width="238" alt="2017-04-07 2 12 50" src="https://cloud.githubusercontent.com/assets/822086/24766824/bff04074-1b37-11e7-8def-5b5a972beb48.png">

### After (customized)

<img width="234" alt="2017-04-07 1 55 59" src="https://cloud.githubusercontent.com/assets/822086/24766770/8af52498-1b37-11e7-946f-e4da61ce20a4.png">

zonuexe commented on php-mode.el:
```diff
@@ -1033,10 +1043,30 @@ PHP heredoc."
 
+(defface php-function-caller-face '((t (:inherit default)))
+  "PHP Mode face used to highlight calling function names."
+  :group 'php-faces)
```
Does anyone have an idea for a better name and description?

Fuco1 replied:
@zonuexe I think I don't understand what it does, so a better description probably is required indeed :P 

Could you post a code snippet to show us the element which would be highlighted with this face?

zonuexe replied:
This face highlights function/method name part of `$obj->method`.

I searched for other implementations, and I found js2-mode have `js2-function-call`. 
I will follow this.

<img width="841" alt="2017-04-07 6 25 36" src="https://cloud.githubusercontent.com/assets/822086/24794195/a9111af8-1bbf-11e7-9d5d-32303bda1fd4.png">

Fuco1 replied:
Oh yes.  I got confused because the *caller* is usually either the object on which you call it or the class in which you use the object/method.

`php-function-call` is good and understandable, but does it work for both functions and methods, or only for methods?  If the latter, I would change it to `php-method-call`.

zonuexe replied:
@Fuco1 Thanks for your review.

I added `php-method-call` face. 
(`php-function-call` and `php-static-method-call` are unused, but I will use them later.)

ejmr commented:
Good work, thanks!  I just have a question before merging:

Why the removal of `-face` from the face names?  Looking at other modes I see that it (apparently) has become a modern practice to not use `-face` for names defined by `defface`.  Whatever the reason, I want to be able to document it in the merge commit message so that future developers years from now will be able to use `git log` to understand *why* we made this change, and not just *what* we changed.

Fuco1 commented:
I've also noticed that you `defvar` the face symbols, that has also become deprecated.

zonuexe commented:
@ejmr @Fuco1 Thank you for reviewing!

I reconstructed commit history of this PR.

> Whatever the reason, I want to be able to document it in the merge commit message so that future developers years from now will be able to use git log to understand why we made this change, and not just what we changed.

I mentioned in https://github.com/ejmr/php-mode/pull/352/commits/4135d680776e92ba477bf895ac8a6ab2afa3a3a2. Is this information sufficient?

> I've also noticed that you `defvar` the face symbols, that has also become deprecated.

I think so, too.  I have deleted it because there is little benefit gained by excess flexibility.

ejmr commented:
Sorry about the slow response this weekend.

@zonuexe That information is definitely sufficient: simple and straight to the point.  Thanks!

@Fuco1 I hope you don't mind me adding a `Reviewed-by` line to the merge commit with your name; it's to acknowledge your helpful feedback.  If you would prefer I omit such lines in the future, simply let me know.

Thanks again to you both!
